File: tools/dump_early_format.py
#!/usr/bin/env python3

import datetime
import hashlib
import re

class ParseError(Exception): pass
class MissingFeatureError(Exception): pass

def dump_main_file(inf, outf):
    helpers = Helpers(inf, outf)
    helpers.read_block_settings()
    inf.seek(0)
    data = inf.read(helpers.datasize + helpers.sumsize + 1)
    if len(data) != helpers.datasize + helpers.sumsize:
        raise ParseError('Main file did not contain exactly 1 data block')
    helpers.dump_settings_block(verify_type=b'ebakup database v1')

def dump_backup_file(inf, outf):
    helpers = Helpers(inf, outf)
    helpers.read_block_settings()
    helpers.dump_settings_block(verify_type=b'ebakup backup data')
    while not helpers.is_final_block_dumped:
        helpers.dump_backup_block()

def dump_content_file(inf, outf):
    helpers = Helpers(inf, outf)
    helpers.read_block_settings()
    helpers.dump_settings_block(verify_type=b'ebakup content data')
    while not helpers.is_final_block_dumped:
        helpers.dump_content_block()

class Helpers(object):
    def __init__(self, inf, outf):
        self.inf = inf
        self.outf = outf
        # Set to true whenever any block beyond end-of-file is
        # attempted dumped:
        self.is_final_block_dumped = False

    def read_block_settings(self):
        self.inf.seek(0)
        data = self.inf.read(10000)
        match = re.search(b'\nedb-blocksize:(\d+)\n', data)
        if not match:
            raise ParseError('Failed to find block size')
        blocksize = int(match.group(1))
        if match.start() >= blocksize:
            raise ParseError('Block size not in first block')
        match = re.search(b'\nedb-blocksum:([^\n]+)\n', data)
        if not match:
            raise ParseError('Failed to find block checksum')
        if match.start() >= blocksize:
            raise ParseError('Block checksum not in first block')
        if match.group(1) == b'sha256':
            self.sumalgo = hashlib.sha256
        else:
            raise MissingFeatureError(
                'Unknown checksum algorithm: ' + match.group(1).decode('utf-8'))
        self.sumsize = self.sumalgo().digest_size
        self.datasize = blocksize - self.sumsize

    def dump_settings_block(self, verify_type=None):
        self.inf.seek(0)
        data = self.inf.read(self.datasize)
        blocksum = self.inf.read(self.sumsize)
        if self.sumalgo(data).digest() != blocksum:
            raise ParseError('Non-matching checksum in settings block')
        end = data.find(b'\x00')
        if end >= 0:
            if data[end:].strip(b'\x00') != b'':
                raise ParseError('Trailing garbage in settings block')
        else:
            end = len(data)
        if data[end-1] != 10: # b'\n'
            raise ParseError('Last entry in first block does not end with LF')
        settings = data[:end-1].split(b'\n')
        if verify_type is not None and settings[0] != verify_type:
            raise ParseError(
                'Wrong type: ' + str(settings[0]) + ' vs ' + str(verify_type))
        self.outf.write(b'type: ' + settings[0] + b'\n')
        for setting in settings[1:]:
            key, value = setting.split(b':', 1)
            self.outf.write(b'setting: ' + key + b':' + value + b'\n')

    def dump_backup_block(self):
        data = self.inf.read(self.datasize)
        blocksum = self.inf.read(self.sumsize)
        if data == b'':
            self.is_final_block_dumped = True
            return
        if len(blocksum) != self.sumsize:
            raise ParseError('Short read (truncated file?)')
        if self.sumalgo(data).digest() != blocksum:
            raise ParseError('Non-matching block checksum')
        done = 0
        while done < len(data):
            if data[done] == 0:
                if data[done:].strip(b'\x00') != b'':
                    raise ParseError('Trailing garbage in backup block')
                return
            elif data[done] == 0x21:
                done += 1
                length, done = _parse_varuint(data, done)
                end = done + length
                kvid, done = _parse_varuint(data, done)
                kv = data[done:end]
                if b'\n' in kv:
                    raise MissingFeatureError(
                        'LF in keys or values is not implemented')
                key, value = kv.split(b':', 1)
                done = end
                self.outf.write(
                    b'key-value: (' + str(kvid).encode('utf-8') + b')' +
                    key + b':' + value + b'\n')
            elif data[done] == 0x22:
                done += 1
                length, done = _parse_varuint(data, done)
                end = done + length
                xid, done = _parse_varuint(data, done)
                kvids = []
                while done < end:
                    kvid, done = _parse_varuint(data, done)
                    kvids.append(kvid)
                if done != end:
                    raise ParseError('kvid overshot item size')
                self.outf.write(
                    b'extradef: (' + str(xid).encode('utf-8') + b')' +
                    str(kvids).encode('utf-8') + b'\n')
            elif data[done] in (0x90, 0x92):
                itemtype = data[done]
                done += 1
                dirid, done = _parse_varuint(data, done)
                parent, done = _parse_varuint(data, done)
                namelen, done = _parse_varuint(data, done)
                name = data[done:done+namelen]
                done += namelen
                if itemtype == 0x92:
                    xid, done = _parse_varuint(data, done)
                self.outf.write(
                    b'dir: (' + str(parent).encode('utf-8') + b'-' +
                    str(dirid).encode('utf-8') + b')')
                if b'\n' in name:
                    raise MissingFeatureError(
                        'LF in file names not implemented')
                self.outf.write(name)
                if itemtype == 0x92:
                    self.outf.write(b'\nextra: ' + str(xid).encode('utf-8'))
                self.outf.write(b'\n')
            elif data[done] in (0x91, 0x93, 0x94):
                itemtype = data[done]
                done += 1
                parent, done = _parse_varuint(data, done)
                namelen, done = _parse_varuint(data, done)
                name = data[done:done+namelen]
                done += namelen
                cidlen, done = _parse_varuint(data, done)
                cid = data[done:done+cidlen]
                done += cidlen
                size, done = _parse_varuint(data, done)
                mtime, nsec, done = _parse_mtime(data, done)
                if itemtype == 0x94:
                    filetype = data[done]
                    done += 1
                if itemtype in (0x93, 0x94):
                    xid, done = _parse_varuint(data, done)
                self.outf.write(b'file: (' + str(parent).encode('utf-8') + b')')
                if b'\n' in name:
                    raise MissingFeatureError(
                        'LF in file names not implemented')
                self.outf.write(name)
                self.outf.write(b'\ncid: ')
                self.outf.write(_hexbytes(cid))
                self.outf.write(
                    b'\nsize: ' + str(size).encode('utf-8') + b'\nmtime: ')
                if mtime.microsecond != nsec // 1000:
                    # Should not happen. _parse_mtime should check this anyway.
                    raise ParseError(
                        'Last-modified time mismatch between '
                        'microsecond and nanosecond')
                self.outf.write(
                    str(mtime.replace(microsecond=0)).encode('utf-8'))
                if nsec != 0:
                    self.outf.write(b'.' + '{:09}'.format(nsec).encode('utf-8'))
                if itemtype == 0x94:
                    self.outf.write(
                        b'\nfile type: ' +
                        self._decode_file_type(filetype).encode('utf-8'))
                if itemtype in (0x93, 0x94):
                    self.outf.write(b'\nextra: ' + str(xid).encode('utf-8'))
                self.outf.write(b'\n')
            else:
                raise ParseError('Unknown data item type: ' + str(data[done]))

    def _decode_file_type(self, filetype):
        if filetype == ord('L'):
            return 'symlink'
        if filetype == ord('S'):
            return 'socket'
        if filetype == ord('P'):
            return 'pipe'
        if filetype == ord('D'):
            return 'device'
        if filetype == ord('?'):
            return 'unknown'
        raise NotImplementedError('Unknown file type: ' + str(filetype))

    def dump_content_block(self):
        data = self.inf.read(self.datasize)
        blocksum = self.inf.read(self.sumsize)
        if data == b'':
            self.is_final_block_dumped = True
            return
        if len(blocksum) != self.sumsize:
            raise ParseError('Short read (truncated file?)')
        if self.sumalgo(data).digest() != blocksum:
            raise ParseError('Non-matching block checksum')
        done = 0
        while done < len(data):
            if data[done] == 0:
                if data[done:].strip(b'\x00') != b'':
                    raise ParseError('Trailing garbage in backup block')
                return
            elif data[done] == 0xdd:
                done += 1
                cidlen, done = _parse_varuint(data, done)
                sumlen, done = _parse_varuint(data, done)
                cslen = max(cidlen, sumlen)
                cs = data[done:done+cslen]
                done += cslen
                first, done = _parse_seconds_after_epoch(data, done)
                last, done = _parse_seconds_after_epoch(data, done)
                self.outf.write(b'cid: ' + _hexbytes(cs[:cidlen]) + b'\n')
                self.outf.write(b'checksum: ')
                if sumlen == cidlen:
                    self.outf.write(b'*')
                else:
                    self.outf.write(_hexbytes(cs[:sumlen]))
                self.outf.write(b'\nfirst: ')
                self.outf.write(
                    _tobytes(first) + b'\nlast: ' + _tobytes(last) + b'\n')
                while data[done] in (0xa0, 0xa1):
                    has_sum = data[done] == 0xa1
                    done += 1
                    checksum = None
                    if has_sum:
                        checksum = data[done:done+sumlen]
                        done += sumlen
                        self.outf.write(b'changed: ' + _hexbytes(checksum))
                    else:
                        self.outf.write(b'restored')
                    first, done = _parse_seconds_after_epoch(data, done)
                    last, done = _parse_seconds_after_epoch(data, done)
                    self.outf.write(b'\nfirst: ')
                    self.outf.write(
                        _tobytes(first) + b'\nlast: ' + _tobytes(last) + b'\n')
            elif data[done] == 0xd0:
                raise NotTestedError('Update entries not implemented yet')
            else:
                raise ParseError('Unknown data item type: ' + str(data[done]))

def _tobytes(obj):
    return str(obj).encode('utf-8')

def _hexbytes(binary):
    return ''.join('{:02x}'.format(x) for x in binary).encode('utf-8')

def _parse_varuint(data, done):
    if data[done] < 0x80:
        return data[done], done+1
    if data[done+1] < 0x80:
        return (data[done] & 0x7f) + (data[done+1] << 7), done+2
    if data[done+2] < 0x80:
        return (data[done] & 0x7f) + ((data[done+1] & 0x7f) << 7) + (data[done+2] << 14), done+3
    value = 0
    shift = 0
    while True:
        value += (data[done] & 0x7f) << shift
        shift += 7
        if data[done] < 0x80:
            return value, done + 1
        done += 1

def _parse_mtime(data, done):
    year = data[done] + (data[done+1] << 8)
    secs = (data[done+2] + (data[done+3] << 8) + (data[done+4] << 16) +
            ((data[done+5] & 0x80) << 17))
    nsecs = ((data[done+5] & 0x3f) + (data[done+6] << 6) +
             (data[done+7] << 14) + (data[done+8] << 22))
    assert year != 0 or (secs == 0 and nsecs == 0)
    assert nsecs >= 0
    assert nsecs < 1000000000
    day = secs // 86400
    assert day >= 0
    assert day < 366
    left = secs - day * 86400
    hour = left // 3600
    assert hour >= 0
    assert hour < 24
    left = left - hour * 3600
    minute = left // 60
    assert minute >= 0
    assert minute < 60
    second = left - minute * 60
    assert second >= 0
    assert second < 60
    month, day = _month_and_day_from_day_of_year(year, day)
    mtime = datetime.datetime(
        year, month, day, hour, minute, second, nsecs//1000)
    return mtime, nsecs, done + 9

daysofmonth = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
def _month_and_day_from_day_of_year(year, day):
    leap_year= year % 400 == 0 or (year % 4 == 0 and year % 100 != 0)
    if not leap_year and day >= 59:
        # skip February 29
        day += 1
    for month, days in enumerate(daysofmonth):
        if day < days:
            return month+1, day+1
        day -= days
    assert False

def _parse_seconds_after_epoch(data, done):
    seconds = (
        data[done] + data[done+1] * 256 +
        data[done+2] * 0x10000 + data[done+3] * 0x1000000)
    return datetime.datetime.utcfromtimestamp(seconds), done + 4

File: tools/dump_early_format.cc
/*
g++ -g -Wall -Wextra -O3 --std=c++11 -o dump_early_format dump_early_format.cc -lgcrypt
 */

#include <iostream>
#include <fstream>
#include <memory>
#include <stdexcept>
#include <string>
#include <vector>

#include <stdint.h>
#include <string.h>

#include <gcrypt.h>

class NotImplementedError: public std::logic_error {
 public:
  explicit NotImplementedError(std::string what) :
      std::logic_error(what) {};
};

class InvalidStateError: public std::logic_error {
 public:
  explicit InvalidStateError(std::string what) :
      std::logic_error(what) {};
};

class CommandLineError: public std::runtime_error {
 public:
  explicit CommandLineError(std::string what) :
      std::runtime_error(what) {};
};

class InvalidDataError: public std::runtime_error {
 public:
  explicit InvalidDataError(std::string what) :
      std::runtime_error(what) {};
};

class Arguments {
 public:
  Arguments(int argc, const char** argv);
  std::ifstream* Input();
  std::ostream* Output();

  std::string infilename;
  std::string outfilename;

 private:
  std::ifstream infile;
  std::ofstream outfile;
};

typedef void (*dumper_t)(std::ifstream&, std::ostream&);
class Main {
 public:
  Main(int argc, const char** argv);
  int Run();

 private:
  void InitializeGCrypt();
  dumper_t GetDumper();
  Arguments args;
};

class Buffer {
 public:
  explicit Buffer(int initial_capacity=0)
      : buffer(nullptr)
      , buf_size(0)
      , buf_capacity(initial_capacity)
      , buf_pos(0) {
    if (buf_capacity > 0)
      buffer = new uint8_t[buf_capacity];
  }

  Buffer& operator=(const Buffer&) = delete;
  Buffer(const Buffer&) = delete;

  ~Buffer() {
    delete[] buffer;
  }

  void Read(std::istream& input, int amt) {
    if (size() + amt > capacity())
      Reserve(size() + amt);
    input.read((char*)(data() + size()), amt);
    buf_size += input.gcount();
  }

  void Reserve(int new_cap) {
    if (new_cap < size())
      return;
    uint8_t * new_data = new uint8_t[new_cap];
    memcpy(new_data, buffer, size());
    delete[] buffer;
    buffer = new_data;
    buf_capacity = new_cap;
  }

  void Resize(int new_size) {
    if (capacity() < new_size)
      Reserve(new_size);
    if (new_size < 0)
      new_size = 0;
    buf_size = new_size;
  }

  int Find(int ch, int start=0, int end=-1) {
    if (start < 0)
      start = 0;
    if (end < 0)
      end = size();
    if (end > size())
      end = size();
    for (int i = start; i < end; i++) {
      if (buffer[i] == ch)
        return i;
    }
    return -1;
  }

  void CheckBufPos() {
    if (buf_pos < 0 || buf_pos > buf_size)
      throw InvalidDataError("Buffer position out of range");
  }
  void Seek(int new_pos) { buf_pos = new_pos; CheckBufPos(); }
  int IsAtEnd() const { return buf_pos >= buf_size; }
  int CurrentOctet() const { return (buffer)[buf_pos]; }
  int ReadOctet() { buf_pos += 1; CheckBufPos(); return (buffer)[buf_pos-1]; }
  long long int ReadVarUint();
  uint8_t * CurrentData() { return buffer + buf_pos; }
  char* CurrentDataChar() { return (char*)(buffer + buf_pos); }
  int CurrentDataSize() { return buf_size - buf_pos; }
  void Skip(int amt) { buf_pos += amt; CheckBufPos(); }
  uint32_t ReadUint32() {
    const uint8_t * d = CurrentData();
    buf_pos += 4;
    return d[0] | (d[1] << 8) | (d[2] << 16) | (d[3] << 24);
  }

  uint8_t * data() { return buffer; };
  int size() { return buf_size; };
  int capacity() { return buf_capacity; };

 private:
  uint8_t * buffer;
  int buf_size;
  int buf_capacity;
  int buf_pos;
};

class Helpers {
 public:
  Helpers(std::ifstream& infile);
  void read_block_settings();
  void dump_settings_block(std::ostream& outfile);
  void dump_next_content_block(std::ostream& outfile);

  int blocksize;
  int blockdatasize;
  int blocksumsize;
  std::string blocksum;
  std::vector<uint8_t> (*blockalgo)(const uint8_t* data, int datalen);

 private:
  std::ifstream& f;

  void WriteDateTimeForSecondsAfterEpoch(
      std::ostream& output, long long int sae);
  void WriteHexEncoded(std::ostream& output, uint8_t* data, int datalen);
  std::unique_ptr<Buffer> ReadNextBlock();
};


int main(int argc, const char ** argv) {
  try {
    Main m(argc, argv);
    return m.Run();
  } catch (CommandLineError& e) {
    std::cout << "ERROR: " << e.what() << "\n";
  } catch (std::runtime_error& e) {
    std::cout << "ERROR: " << e.what() << "\n";
  }
}

/* --------------------------------------------------
 * Helper functions
 * --------------------------------------------------
 */

int string_to_int(char* str, int strsize) {
  int value = 0;
  for (int i = 0; i < strsize; i++) {
    if (str[i] < '0' || str[i] > '9')
      throw std::runtime_error(
          "Could not parse string as value: " + std::string(str, strsize));
    value *= 10;
    value += str[i] - '0';
  }
  return value;
}

std::vector<uint8_t> calculate_unknown_checksum(
    const uint8_t* /*data*/, int /*datalen*/) {
  throw InvalidStateError("Checksum algorithm not initialized");
}

std::vector<uint8_t> calculate_sha256(const uint8_t* data, int datalen) {
  if (gcry_md_get_algo_dlen(GCRY_MD_SHA256) != 32)
    throw std::runtime_error(
        "gcrypt says sha256 is " +
        std::to_string(gcry_md_get_algo_dlen(GCRY_MD_SHA256)) + " bytes long!");
  gcry_md_hd_t hd;
  gcry_error_t err = gcry_md_open(&hd, GCRY_MD_SHA256, 0);
  if (gcry_err_code(err) != GPG_ERR_NO_ERROR)
    throw std::runtime_error("gcry_md_open returned an error");
  gcry_md_write(hd, (char*)data, datalen);
  unsigned char* digest = gcry_md_read(hd, GCRY_MD_SHA256);
  std::vector<uint8_t> digestvec((uint8_t*)digest, (uint8_t*)(digest+32));
  gcry_md_close(hd);
  return digestvec;
}

/* --------------------------------------------------
 * dumper for content
 * --------------------------------------------------
 */

void dump_content_file(std::ifstream& infile, std::ostream& outfile) {
  Helpers helpers(infile);
  helpers.read_block_settings();
  helpers.dump_settings_block(outfile);
  while (!infile.eof()) {
    if (!infile.good())
      throw std::runtime_error("Input file not in good state!");
    helpers.dump_next_content_block(outfile);
  }
}

/* --------------------------------------------------
 * class Arguments
 * --------------------------------------------------
 */

Arguments::Arguments(int argc, const char** argv) {
  int arg = 1;
  while (arg < argc) {
    std::string argstr(argv[arg]);
    if (argstr == "-o" || argstr == "--output") {
      if (arg + 1 >= argc)
        throw CommandLineError("Option '" + argstr + "' requires an argument");
      std::string ofn(argv[arg+1]);
      if (!outfilename.empty())
        throw CommandLineError(
            "Output file name set twice: " + outfilename + " and " + ofn);
      outfilename = ofn;
      arg += 2;
    } else if (argstr[0] == '-') {
      throw CommandLineError("Unknown option: " + argstr);
    } else {
      if (!infilename.empty())
        throw CommandLineError(
            "Input file name set twice: " + infilename + " and " + argstr);
      infilename = argstr;
      arg += 1;
    }
  }

  if (infilename.empty())
    throw CommandLineError("Required argument missing: input file name");
}

std::ifstream* Arguments::Input() {
  if (infile.is_open())
    return &infile;
  infile.open(infilename, std::ios_base::in | std::ios_base::binary);
  return &infile;
}

std::ostream* Arguments::Output() {
  if (outfile.is_open())
    return &outfile;
  if (outfilename.empty())
    return &std::cout;
  outfile.open(
      outfilename,
      std::ios_base::out | std::ios_base::trunc | std::ios_base::binary);
  return &outfile;
}

/* --------------------------------------------------
 * class Main
 * --------------------------------------------------
 */

Main::Main(int argc, const char** argv)
    : args(argc, argv) {
  InitializeGCrypt();
}

void Main::InitializeGCrypt() {
  // Mostly Copied from the gcrypt documentation.
  /* Version check should be the very first call because it
     makes sure that important subsystems are intialized. */
  if (!gcry_check_version (GCRYPT_VERSION))
    throw std::runtime_error("libgcrypt version mismatch\n");

  /* Disable secure memory.  */
  gcry_control (GCRYCTL_DISABLE_SECMEM, 0);

  /* Tell Libgcrypt that initialization has completed. */
  gcry_control (GCRYCTL_INITIALIZATION_FINISHED, 0);
}

int Main::Run() {
  dumper_t dumper = GetDumper();
  std::ostream* output = args.Output();
  *output << "event: dump start\n";
  dumper(*args.Input(), *output);
  *output << "event: dump complete\n";
  return 1;
}

dumper_t Main::GetDumper() {
  std::ifstream* infile = args.Input();
  infile->seekg(0);
  char buf[100];
  infile->read(buf, 100);
  int bufsize = infile->gcount();
  std::string start(buf, bufsize);
  if (start.find("ebakup content data\n") == 0)
    return dump_content_file;
  throw std::runtime_error(
      "Failed to recognize the file type of the input file (" +
      args.infilename + ")");
}

/* --------------------------------------------------
 * class Buffer
 * --------------------------------------------------
 */

long long int Buffer::ReadVarUint() {
  long long int value = 0;
  while (buf_pos < buf_size) {
    value <<= 7;
    value |= buffer[buf_pos] & 0x7f;
    buf_pos += 1;
    if (buffer[buf_pos-1] < 0x80)
      return value;
  }
  throw InvalidDataError("Varuint didn't end before the buffer");
}

/* --------------------------------------------------
 * class Helpers
 * --------------------------------------------------
 */
Helpers::Helpers(std::ifstream& infile)
    : blocksize(0)
    , blockdatasize(0)
    , blocksumsize(0)
    , blockalgo(calculate_unknown_checksum)
    , f(infile) {
}

void Helpers::read_block_settings() {
  char buf[10000];
  f.seekg(0);
  f.read(buf, 10000);
  int bufsize = f.gcount();
  char* sizestr = (char*)memmem(buf, bufsize, "\nedb-blocksize:", 15);
  if (!sizestr)
    throw InvalidDataError("No blocksize specified in data file");
  sizestr += 15;
  char* sizeend = (char*)memchr(sizestr, '\n', bufsize - (sizestr - buf));
  if (!sizeend)
    throw InvalidDataError("Failed to find end of blocksize value");
  blocksize = string_to_int(sizestr, sizeend - sizestr);
  if (sizeend - buf > blocksize)
    throw InvalidDataError("No blocksize specified in settings block");
  char* sumstr = (char*)memmem(buf, bufsize, "\nedb-blocksum:", 14);
  if (!sumstr || (sumstr - buf) > blocksize)
    throw InvalidDataError("No block checksum specified in settings block");
  sumstr += 14;
  char* sumend = (char*)memchr(sumstr, '\n', bufsize - (sumstr - buf));
  if (!sumend)
    throw InvalidDataError("Failed to find end of block checksum value");
  blocksum = std::string(sumstr, sumend - sumstr);
  if (blocksum == "sha256") {
    blocksumsize = 32;
    blockalgo = calculate_sha256;
  } else
    throw NotImplementedError("Unknown block checksum: " + blocksum);
  blockdatasize = blocksize - blocksumsize;
}

void Helpers::dump_settings_block(std::ostream& output) {
  f.seekg(0);
  std::unique_ptr<Buffer> block = ReadNextBlock();
  int done = 0;
  uint8_t * data = block->data();
  while (done < block->size()) {
    if (data[done] == 0) {
      while (done < block->size()) {
        if (data[done] != 0)
          throw InvalidDataError("Trailing garbage in settings block");
        done += 1;
      }
      return;
    }
    int end = block->Find('\n', done);
    if (end < 0)
      throw InvalidDataError("Failed to find end of setting");
    if (done == 0)
      output << "type: ";
    else {
      if (block->Find(':', done, end) < 0)
        throw InvalidDataError("No ':' in setting line");
      output << "setting: ";
    }
    output.write((char*)data + done, end - done + 1);
    done = end + 1;
  }
}

void Helpers::dump_next_content_block(std::ostream& output) {
  std::unique_ptr<Buffer> block = ReadNextBlock();
  block->Seek(0);
  while (!block->IsAtEnd()) {
    if (block->CurrentOctet() == 0xdd) {
      block->Skip(1);
      int cidlen = block->ReadVarUint();
      int sumlen = block->ReadVarUint();
      int cslen = std::max(cidlen, sumlen);
      output << "cid: ";
      WriteHexEncoded(output, block->CurrentData(), cidlen);
      output << "\nchecksum: ";
      if (cidlen == sumlen)
        output << "*";
      else
        WriteHexEncoded(output, block->CurrentData(), sumlen);
      output << "\n";
      block->Skip(cslen);
      long long int first = block->ReadUint32();
      long long int last = block->ReadUint32();
      output << "first: ";
      WriteDateTimeForSecondsAfterEpoch(output, first);
      output << "\nlast: ";
      WriteDateTimeForSecondsAfterEpoch(output, last);
      output << "\n";
      while (block->CurrentOctet() == 0xa0 || block->CurrentOctet() == 0xa1) {
        if (block->CurrentOctet() == 0xa1) {
          output << "changed: ";
          WriteHexEncoded(output, block->CurrentData(), sumlen);
          output << "\n";
        } else
          output << "restored\n";
        first = block->ReadUint32();
        last = block->ReadUint32();
        output << "first: ";
        WriteDateTimeForSecondsAfterEpoch(output, first);
        output << "\nlast: ";
        WriteDateTimeForSecondsAfterEpoch(output, last);
        output << "\n";
      }
    } else if (block->CurrentOctet() == 0) {
      while (!block->IsAtEnd()) {
        if (block->ReadOctet() != 0)
          throw InvalidDataError("Trailing garbage in content block");
      }
      return;
    } else
      throw InvalidDataError(
          "Unknown data type: " + std::to_string(block->CurrentOctet()));
  }
}

const int seconds_per_minute = 60;
const int seconds_per_hour = seconds_per_minute * 60;
const int seconds_per_day = seconds_per_hour * 24;
const long long int seconds_per_year = seconds_per_day * 365;
const int days_per_month[] = { 31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 };
//const int seconds_per_leapyear = seconds_per_day * 366;
void Helpers::WriteDateTimeForSecondsAfterEpoch(
    std::ostream& output, long long int sae) {
  if (sae < 0)
    throw NotImplementedError("Negative time stamps are not correctly handled");
  int years_no_leap = sae / seconds_per_year;
  long long int left = sae - years_no_leap * seconds_per_year;
  int leap_years = (years_no_leap + 2) / 4;
  leap_years -= (years_no_leap + 70) / 100;
  leap_years += (years_no_leap + 370) / 400;
  while (leap_years * seconds_per_day > left) {
    years_no_leap -= 1;
    left = sae - years_no_leap * seconds_per_year;
    leap_years = (years_no_leap + 2) / 4;
    leap_years -= (years_no_leap + 70) / 100;
    leap_years += (years_no_leap + 370) / 400;
  }
  int year = 1970 + years_no_leap;
  left -= leap_years * seconds_per_day;
  int days = left / seconds_per_day;
  left -= days * seconds_per_day;
  int day = days + 1;
  if (day > 59) {
    if (!(year % 400 == 0 || (year % 4 == 0 && year % 100 != 0)))
      day += 1;
  }
  int month = 0;
  int day_of_month = day;
  while (month < 11 && day_of_month > days_per_month[month]) {
    day_of_month -= days_per_month[month];
    month += 1;
  }
  month += 1;
  int hour = left / seconds_per_hour;
  left -= hour * seconds_per_hour;
  int minute = left / seconds_per_minute;
  int second = left - minute * seconds_per_minute;
  const char* ymsep = month < 10 ? "-0" : "-";
  const char* mdsep = day_of_month < 10 ? "-0" : "-";
  const char* dhsep = hour < 10 ? " 0" : " ";
  const char* hmsep = minute < 10 ? ":0" : ":";
  const char* mssep = second < 10 ? ":0" : ":";
  output << year << ymsep << month << mdsep << day_of_month << dhsep <<
      hour << hmsep << minute << mssep << second;
}

const char hexits[] = {
  '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
  'a', 'b', 'c', 'd', 'e', 'f' };
void Helpers::WriteHexEncoded(
    std::ostream& output, uint8_t* data, int datalen) {
  for (int i = 0; i < datalen; i++) {
    char n[2];
    n[0] = hexits[(data[i] >> 4) & 15];
    n[1] = hexits[data[i] & 15];
    output.write(n, 2);
  }
}

std::unique_ptr<Buffer> Helpers::ReadNextBlock() {
  std::unique_ptr<Buffer> buf(new Buffer);
  buf->Read(f, blocksize);
  if (buf->size() == 0) {
    if (f.eof())
      return buf;
    throw std::runtime_error(
        "Got no data even though it is not the end of the file");
  }
  if (buf->size() < blocksize) {
    throw std::runtime_error(
        "Got incomplete block (" + std::to_string(buf->size()) +
        " octets instead of " + std::to_string(blocksize) + ")\n");
  }
  if (buf->size() != blocksize) {
    throw std::runtime_error(
        "Got wrong amount of block data (" + std::to_string(buf->size()) +
        " octets instead of " + std::to_string(blocksize) + ")\n");
  }
  std::vector<uint8_t> checksum = blockalgo(buf->data(), blockdatasize);
  if (checksum.size() != (unsigned int)blocksumsize)
    throw std::runtime_error(
        "Block checksum did not have expected size (" +
        std::to_string(checksum.size()) + " vs " +
        std::to_string(blocksumsize));
  if (memcmp(checksum.data(), buf->data() + blockdatasize, blocksumsize) != 0)
    throw std::runtime_error("Block checksum mismatch!");
  buf->Resize(blockdatasize);
  return buf;
}
